### PR TITLE
Replace EXTRACT-specific code with code for flat subtree handling

### PR DESCRIPTION
**Pull request by tadhg-ohiggins**
Add FlatsubtreeMatcher and FlatParagraphProcessor classes to deal with elements that need to have their sub-elements processed into node trees rather than completely flattened; currently EXTRACT and EXAMPLE are the elements we're concered with. These replace ExtractMatcher and ExtractPargraphProcessor.

### CODE REVIEW

File: regparser/tree/xml_parser/flatsubtree_processor.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+    def matches(self, xml):
+        return xml.tag in self.tags
+
+    def derive_nodes(self, xml, processor=None):
+        processor = FlatParagraphProcessor()
+        text = (xml.text or '').strip()
+        extract_node = Node(text=text, node_type=self.node_type,
+                            label=[mtypes.MARKERLESS])
+        return [processor.process(xml, extract_node)]
```
Comment (cmc333333) [inline regparser/tree/xml_parser/flatsubtree_processor.py]: s/`extract_node`/`root_node` or even just `node`?

File: regparser/tree/xml_parser/flatsubtree_processor.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+                paragraph_processor.HeaderMatcher(),
+                paragraph_processor.SimpleTagMatcher('P', 'FP')]
+
+
+class FlatsubtreeMatcher(paragraph_processor.BaseMatcher):
+    """
+    Detects tags passed to it on init and processes them with the
+    FlatParagraphProcessor. Also optionally sets node_type.
+    """
+    def __init__(self, tags=[], node_type="regtext"):
```
Comment (cmc333333) [inline regparser/tree/xml_parser/flatsubtree_processor.py]: What do you think about using the constant, `Node.REGTEXT`?

File: regparser/tree/xml_parser/reg_text.py [hunk 3]
```diff
@@ -292,7 +293,9 @@ class RegtextParagraphProcessor(paragraph_processor.ParagraphProcessor):
     MATCHERS = [paragraph_processor.StarsMatcher(),
                 paragraph_processor.TableMatcher(),
                 paragraph_processor.FencedMatcher(),
-                extract_processor.ExtractMatcher(),
+                flatsubtree_processor.FlatsubtreeMatcher(tags=['EXTRACT'],
+                                                         node_type="extract"),
```
Comment (cmc333333) [inline regparser/tree/xml_parser/reg_text.py]: Similar to the above, would it make more sense to use the constant, `Node.EXTRACT`?

File: regparser/tree/xml_parser/flatsubtree_processor.py (hunk 1 above)
Comment (tadhg-ohiggins) [inline regparser/tree/xml_parser/flatsubtree_processor.py]: Should've been `node`.

File: regparser/tree/xml_parser/flatsubtree_processor.py (hunk 2 above)
Comment (tadhg-ohiggins) [inline regparser/tree/xml_parser/flatsubtree_processor.py]: That's the better approach.

File: regparser/tree/xml_parser/reg_text.py (hunk 3 above)
Comment (tadhg-ohiggins) [inline regparser/tree/xml_parser/reg_text.py]: Yup.

**Comment by cmc333333** [comment]
:+1:

**Pull request merged**